File: src/openfl/Memory.nim
import ./utils/ByteArray

type
  Memory* = ref object

proc select*(bytes: ByteArray) =
  # TODO: implement byte-level memory access
  discard

proc getByte*(addr: int): int = 0
proc getDouble*(addr: int): float64 = 0.0
proc getFloat*(addr: int): float32 = 0.0
proc getInt*(addr: int): int = 0
proc getShort*(addr: int): int = 0

proc setByte*(addr: int, v: int) = discard
proc setDouble*(addr: int, v: float64) = discard
proc setFloat*(addr: int, v: float32) = discard
proc setInt*(addr: int, v: int) = discard
proc setShort*(addr: int, v: int) = discard

File: kfl/src/main/kotlin/openfl/Memory.kt
package openfl

import openfl.utils.ByteArray

object Memory {
    private var __byteArray: ByteArray? = null

    fun select(ba: ByteArray?) {
        __byteArray = ba
    }

    fun getByte(addr: Int): Int {
        val pos = __byteArray?.position ?: 0
        __byteArray?.position = addr
        val result = __byteArray?.readByte() ?: 0
        __byteArray?.position = pos
        return result
    }

    fun setByte(addr: Int, v: Int) {
        val pos = __byteArray?.position ?: 0
        __byteArray?.position = addr
        __byteArray?.writeByte(v)
        __byteArray?.position = pos
    }

    fun getInt(addr: Int): Int {
        val pos = __byteArray?.position ?: 0
        __byteArray?.position = addr
        val result = __byteArray?.readInt() ?: 0
        __byteArray?.position = pos
        return result
    }

    fun setInt(addr: Int, v: Int) {
        val pos = __byteArray?.position ?: 0
        __byteArray?.position = addr
        __byteArray?.writeInt(v)
        __byteArray?.position = pos
    }

    fun getDouble(addr: Int): Double {
        val pos = __byteArray?.position ?: 0
        __byteArray?.position = addr
        val result = __byteArray?.readDouble() ?: 0.0
        __byteArray?.position = pos
        return result
    }

    fun setDouble(addr: Int, v: Double) {
        val pos = __byteArray?.position ?: 0
        __byteArray?.position = addr
        __byteArray?.writeDouble(v)
        __byteArray?.position = pos
    }
}
